File: solutions/code-carl/ch05_stack/t3_lc0020_is_valid.py
"""LeetCode 20. Valid Parentheses
https://leetcode.cn/problems/valid-parentheses/
"""


class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        dic = {'(': ')', '[': ']', '{': '}'}
        for ch in s:
            if ch in dic:
                stack.append(dic[ch])
            else:
                if not stack or ch != stack.pop():
                    return False
        return not stack


if __name__ == "__main__":
    solution = Solution()
    assert solution.isValid("()") is True
    assert solution.isValid("()[]{}") is True
    assert solution.isValid("(]") is False
    print("All tests passed.")

File: solutions/code-carl/ch05_stack/T3_LC0020_IsValid.java
/**
 * Practice template for:
 * - LeetCode 20. Valid Parentheses
 *   https://leetcode.cn/problems/valid-parentheses/
 */
import java.util.ArrayDeque;
import java.util.Deque;

public class T3_LC0020_IsValid {
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        char[] input = s.toCharArray();
        for (char ch : input) {
            if (ch == '(') {
                stack.push(')');
            } else if (ch == '[') {
                stack.push(']');
            } else if (ch == '{') {
                stack.push('}');
            } else if (stack.isEmpty() || stack.pop() != ch) {
                return false;
            }
        }
        return stack.isEmpty();
    }
}
